File: backend/src/test/java/com/rungo/api/domain/notification/listener/NotificationEventListenerTest.java
package com.rungo.api.domain.notification.listener;

import com.rungo.api.domain.notification.event.MarathonCanceledEvent;
import com.rungo.api.domain.notification.event.RegistrationCompletedEvent;
import com.rungo.api.domain.notification.support.NotificationEmailFactory;
import com.rungo.api.global.infrastructure.mail.EmailMessage;
import com.rungo.api.global.infrastructure.mail.EmailService;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@ExtendWith(MockitoExtension.class)
class NotificationEventListenerTest {

    @Mock
    private EmailService emailService;

    @Mock
    private NotificationEmailFactory notificationEmailFactory;

    @InjectMocks
    private NotificationEventListener notificationEventListener;

    @Test
    @DisplayName("접수 완료 이벤트 발생 시 이메일 메시지를 생성하고 발송한다")
    void handleRegistrationCompleted_test() {
        RegistrationCompletedEvent event =
                new RegistrationCompletedEvent("test@gmail.com", "서울 마라톤", "10km");

        EmailMessage message = new EmailMessage(
                "test@gmail.com",
                "[Rungo] 서울 마라톤 참가 접수 완료 안내",
                "body"
        );

        given(notificationEmailFactory.registrationCompleted(
                event.email(), event.marathonTitle(), event.courseName())
        ).willReturn(message);

        notificationEventListener.handleRegistrationCompleted(event);

        verify(notificationEmailFactory, times(1))
                .registrationCompleted(event.email(), event.marathonTitle(), event.courseName());
        verify(emailService, times(1)).send(message);
    }

    @Test
    @DisplayName("대회 취소 이벤트 발생 시 참가자 수만큼 이메일 메시지를 생성하고 발송한다")
    void handleMarathonCanceled_test() {
        List<String> emails = List.of("user1@gmail.com", "user2@gmail.com");
        MarathonCanceledEvent event = new MarathonCanceledEvent("서울 마라톤", emails);

        given(notificationEmailFactory.marathonCanceled("user1@gmail.com", "서울 마라톤"))
                .willReturn(new EmailMessage("user1@gmail.com", "subject1", "body1"));
        given(notificationEmailFactory.marathonCanceled("user2@gmail.com", "서울 마라톤"))
                .willReturn(new EmailMessage("user2@gmail.com", "subject2", "body2"));

        notificationEventListener.handleMarathonCanceled(event);

        verify(notificationEmailFactory, times(1)).marathonCanceled("user1@gmail.com", "서울 마라톤");
        verify(notificationEmailFactory, times(1)).marathonCanceled("user2@gmail.com", "서울 마라톤");
        verify(emailService, times(2)).send(org.mockito.ArgumentMatchers.any(EmailMessage.class));
    }
}

File: backend/src/test/java/com/rungo/api/domain/notification/listener/NotificationEventListenerTest.kt
package com.rungo.api.domain.notification.listener

import com.rungo.api.domain.notification.event.MarathonCanceledEvent
import com.rungo.api.domain.notification.event.RegistrationCompletedEvent
import com.rungo.api.domain.notification.support.NotificationEmailFactory
import com.rungo.api.global.infrastructure.mail.EmailMessage
import com.rungo.api.global.infrastructure.mail.entity.EmailOutbox
import com.rungo.api.global.infrastructure.mail.entity.EmailOutboxStatus
import com.rungo.api.global.infrastructure.mail.repository.EmailOutboxRepository
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentCaptor
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class NotificationEventListenerTest {

    @InjectMocks
    private lateinit var notificationEventListener: NotificationEventListener

    @Mock
    private lateinit var notificationEmailFactory: NotificationEmailFactory

    @Mock
    private lateinit var emailOutboxRepository: EmailOutboxRepository

    @Test
    @DisplayName("접수 완료 이벤트가 발행되면 이메일 발송 정보를 Outbox에 저장한다")
    fun handleRegistrationCompleted_save_outbox() {
        val event = RegistrationCompletedEvent(
            email = "user@test.com",
            marathonTitle = "서울 마라톤",
            courseName = "10K"
        )

        given(
            notificationEmailFactory.registrationCompleted(
                event.email,
                event.marathonTitle,
                event.courseName
            )
        ).willReturn(
            EmailMessage(
                to = "user@test.com",
                subject = "[Rungo] 참가 접수 완료",
                body = "서울 마라톤 10K 접수가 완료되었습니다."
            )
        )

        notificationEventListener.handleRegistrationCompleted(event)

        val captor = ArgumentCaptor.forClass(EmailOutbox::class.java)

        verify(emailOutboxRepository).save(captureEmailOutbox(captor))

        val savedOutbox = captor.value
        assertEquals("user@test.com", savedOutbox.recipient)
        assertEquals("[Rungo] 참가 접수 완료", savedOutbox.subject)
        assertEquals("서울 마라톤 10K 접수가 완료되었습니다.", savedOutbox.body)
        assertEquals(EmailOutboxStatus.PENDING, savedOutbox.status)
        assertEquals(0, savedOutbox.retryCount)
    }

    @Test
    @DisplayName("대회 취소 이벤트가 발행되면 참가자 수만큼 Outbox에 저장한다")
    fun handleMarathonCanceled_save_outboxes() {
        val event = MarathonCanceledEvent(
            marathonTitle = "서울 마라톤",
            participantEmails = listOf("user1@test.com", "user2@test.com")
        )

        given(
            notificationEmailFactory.marathonCanceled(
                "user1@test.com",
                event.marathonTitle
            )
        ).willReturn(
            EmailMessage(
                to = "user1@test.com",
                subject = "[Rungo] 대회 취소 안내",
                body = "서울 마라톤 대회가 취소되었습니다."
            )
        )

        given(
            notificationEmailFactory.marathonCanceled(
                "user2@test.com",
                event.marathonTitle
            )
        ).willReturn(
            EmailMessage(
                to = "user2@test.com",
                subject = "[Rungo] 대회 취소 안내",
                body = "서울 마라톤 대회가 취소되었습니다."
            )
        )

        notificationEventListener.handleMarathonCanceled(event)

        val captor = ArgumentCaptor.forClass(EmailOutbox::class.java)

        verify(emailOutboxRepository, times(2)).save(captureEmailOutbox(captor))

        val savedOutboxes = captor.allValues

        assertEquals(2, savedOutboxes.size)

        assertEquals("user1@test.com", savedOutboxes[0].recipient)
        assertEquals("[Rungo] 대회 취소 안내", savedOutboxes[0].subject)
        assertEquals("서울 마라톤 대회가 취소되었습니다.", savedOutboxes[0].body)
        assertEquals(EmailOutboxStatus.PENDING, savedOutboxes[0].status)
        assertEquals(0, savedOutboxes[0].retryCount)

        assertEquals("user2@test.com", savedOutboxes[1].recipient)
        assertEquals("[Rungo] 대회 취소 안내", savedOutboxes[1].subject)
        assertEquals("서울 마라톤 대회가 취소되었습니다.", savedOutboxes[1].body)
        assertEquals(EmailOutboxStatus.PENDING, savedOutboxes[1].status)
        assertEquals(0, savedOutboxes[1].retryCount)
    }

    private fun captureEmailOutbox(captor: ArgumentCaptor<EmailOutbox>): EmailOutbox {
        captor.capture()
        return EmailOutbox.create("", "", "")
    }
}